components/web_discovery/browser/credential_manager.cc

```
@@ -0,0 +1,350 @@
+      sender: "Brave Web Discovery HPNv2 Join"
+      description:
+        "Retrieves anonymous credentials in order to sign Web Discovery
+         measurements sent via the HumanWeb Proxy Network."
+      trigger:
+        "Requests are automatically sent on daily intervals "
+        "while Brave is running."
+      data: "Configuration attributes"
+      destination: WEBSITE
+    }
+    policy {
+      cookies_allowed: NO
+      setting:
+        "Users can opt-in or out via brave://settings/search"
+    })");
+
+}  // namespace
+
+CredentialManager::CredentialManager(
+    PrefService* profile_prefs,
+    network::SharedURLLoaderFactory* shared_url_loader_factory,
+    const ServerConfigLoader* server_config_loader)
+    : profile_prefs_(profile_prefs),
+      shared_url_loader_factory_(shared_url_loader_factory),
+      server_config_loader_(server_config_loader),
+      join_url_(GetDirectHPNHost() + kJoinPath),
+      backoff_entry_(&kBackoffPolicy),
+      background_task_runner_(base::ThreadPool::CreateSequencedTaskRunner({})),
+      background_credential_helper_(background_task_runner_,
+                                    BackgroundCredentialHelper::Create()) {}
+
+CredentialManager::~CredentialManager() = default;
+
+bool CredentialManager::LoadRSAKey() {
+  std::string private_key_b64 =
+      profile_prefs_->GetString(kCredentialRSAPrivateKey);
+
+  if (private_key_b64.empty()) {
+    return true;
+  }
+
+  auto key_info = ImportRSAKeyPair(private_key_b64);
+  if (!key_info) {
+    VLOG(1) << "Failed to import stored RSA key";
+    return false;
+  }
+
+  CHECK(key_info->private_key);
+  rsa_public_key_b64_ = std::move(key_info->public_key_b64);
+  background_credential_helper_
+      .AsyncCall(&BackgroundCredentialHelper::SetRSAKey)
+      .WithArgs(std::move(key_info->private_key));
+
+  return true;
+}
+
+void CredentialManager::OnNewRSAKey(std::unique_ptr<RSAKeyInfo> key_info) {
+  if (!key_info) {
+    VLOG(1) << "RSA key generation failed";
```

shouldn't this report an error back?

---

this error is very unlikely, so a silent failure seems appropriate here

---

what would cause this?

---

mostly invalid input to underlying boringssl functions

---

I mean what's the end result in terms of functionality?

---

also when you say "very unlikely", what circumstances would lead to this? Can you document that with comments?

---

also debug logs should not be built in production code and should generally be removed before merge or changed to DVLOG, but this log doesn't seem helpful if it's a rare case and won't show up unless you are already running with vlog enabled

---

>I mean what's the end result in terms of functionality?

Web Discovery would cease to function until a browser restart. Documented the error case in a comment